programs/ssl/ssl_server2.c

```diff
@@ -3332,6 +3349,16 @@ int main( int argc, char *argv[] )
     }
 #endif
 
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+#if defined(MBEDTLS_SSL_SERVER_NAME_INDICATION)
+#if defined(MBEDTLS_KEY_EXCHANGE_CERT_REQ_ALLOWED_ENABLED)
+    /* exercise setting DN hints for server certificate request */
+    if( opt.cert_req_dn_hint == 3 && key_cert_init2 )
+        mbedtls_ssl_set_hs_dn_hints( &ssl, &srvcert2 );
```

Comment:
Hm, I'm not comfortable calling a `set_hs` function from main: this is not guaranteed to work, and perhaps more importantly, it testing it this way rather than the way it will be used in the real work provides less confidence that it will work when used in the normal way. Do you think you could move that to one of the callbacks (SNI or cert callback)?

Reply:
Why not guaranteed to work?  It is called after the connection is made and the handshake has been initialized.  Putting it in the SNI callback requires SNI.  Putting it in the cert callback is possible, but I would also need to access the file-scoped global 'opt.cert_req_dn_hint' to test configuring it.

> I think it would make sense for them to be present even when SNI is not compiled in.
> However, this is pre-existing, so you can either fix it in this PR, or create an issue to track it independently.

I am hoping this PR will be included in mbedtls 3.2.0, so I think filing a separate issue is the place to consider expanding access.  Personally, I think that the dependency on SNI make sense.  I think that without SNI used to make a different choice, you might as well configure these in `set_conf` rather than in `set_hs`.

Reply:
I expanded the comment.

If I moved this to the cert_callback, I would not have access to `&srvcert2` (defined and scoped in main()) without further work.  Also, the `cert_callback` currently uses `user_data` to store the SNI information, so that would have to be reworked and expanded to use a custom struct with additional members.  While possible, I hope to convince you that this should be addressed in a future issue after determining how much `set_hs` is exposed without SNI (e.g. when `!defined(MBEDTLS_SSL_SERVER_NAME_INDICATION)`)
:rocket:

Reply:
> I am hoping this PR will be included in mbedtls 3.2.0, so I think filing a separate issue is the place to consider expanding access.

Agreed. Also, thanks for explaining why moving to a callback wouldn't be convenient, I agree we can skip it for now and revisit later when we address the issue about which guards.

(By not guaranteed to work, I mean that the documentation says `set_hs` function are for use in callbacks, it doesn't say anything about use outside. You know it works because you're familiar with the code and know when the handshake structure is initialized, but that's implementation details that could change in a minor version. We do not consider this part of the contract.)